### PR TITLE
unix: refactor uv__fs_copyfile() logic

### PR DESCRIPTION
**Pull request by cjihrig**
The Unix/macOS `uv__fs_copyfile()` implementation falls back to using `uv_fs_sendfile()`. This commit refactors the error handling to use the `sendfile()` `req`'s `result` field, which is a `ssize_t` instead of using the return value, which is an `int`. The `int` value was coming back as a negative number for large files on some machines.

Refs: https://github.com/nodejs/node/issues/30085
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1620/

EDIT: The CI has no new failures. I also ran Node's copyfile tests, as well as verifying that this fixed the linked Node issue.

### CODE REVIEW

File: src/unix/fs.c [hunk 1]
```diff
@@ -1076,16 +1076,12 @@ static ssize_t uv__fs_copyfile(uv_fs_t* req) {
...
-                         in_offset,
-                         bytes_to_send,
-                         NULL);
+    uv_fs_sendfile(NULL, &fs_req, dstfd, srcfd, in_offset, bytes_to_send, NULL);
     uv_fs_req_cleanup(&fs_req);
-    if (err < 0)
+    if (fs_req.result < 0) {
+      err = fs_req.result;
       break;
+    }
```
Comment (santigimeno) [inline src/unix/fs.c]: Style nit: extra line after `}`?

**Comment by saghul** [comment]
Not sure I understand how / why this works. You are still assigning the result field to `err` which is an `int`, so isn't it the same? When no callback is passed that's what the function will return anyway: https://github.com/libuv/libuv/blob/a62f8ced7af51d1c4fd8abceb521bd24f362ab14/src/unix/fs.c#L153 Am I missing something?

File: src/unix/fs.c [hunk 2]
```diff
@@ -1076,16 +1076,12 @@ static ssize_t uv__fs_copyfile(uv_fs_t* req) {
...
-                         &fs_req,
-                         dstfd,
-                         srcfd,
-                         in_offset,
-                         bytes_to_send,
-                         NULL);
+    uv_fs_sendfile(NULL, &fs_req, dstfd, srcfd, in_offset, bytes_to_send, NULL);
     uv_fs_req_cleanup(&fs_req);
-    if (err < 0)
+    if (fs_req.result < 0) {
```
Comment (saghul) [inline src/unix/fs.c]: Here we are accessing `fs_req` after cleaning it up. It doesn't seem like a safe thing to do, we might `memset` the thing to 0 or anything in debug mode, for example. (Not saying we do it now, but we might decide to do that one day)

**Comment by santigimeno** [comment]
> Not sure I understand how / why this works. You are still assigning the result field to err which is an int, so isn't it the same?

But you're assigning it after the `if (fs_req.result < 0) {` so shouldn't it be safe?

> When no callback is passed that's what the function will return anyway

Yeah, in the sync version we can have false negatives. Shouldn't we do something like:
```c
return req->result < 0 ? req->result : 0;
```
and that the user just checks the result by using the `uv_fs_get_result()` getter.

**Comment by cjihrig** [comment]
Rerun of the CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1621/. Still no new failures.

**Pull request merged**